src/commands/react-native/upload.ts

```diff
@@ -0,0 +1,319 @@
+  private addRepositoryDataToPayloads = async (payloads: RNSourcemap[]) => {
+    const repositoryData = await getRepositoryData(await newSimpleGit(), this.context.stdout, this.repositoryURL)
+    if (repositoryData === undefined) {
+      return
+    }
+    await Promise.all(
+      payloads.map(async (payload) => {
+        const repositoryPayload = this.getRepositoryPayload(repositoryData!, payload.sourcemapPath)
+        payload.addRepositoryData({
+          gitCommitSha: repositoryData.hash,
+          gitRepositoryPayload: repositoryPayload,
+          gitRepositoryURL: repositoryData.remote,
+        })
+      })
+    )
+  }
+
+  // Looks for the sourcemaps and minified files on disk and returns
+  // the associated payloads.
+  private getMatchingRNSourcemapFiles = async (): Promise<RNSourcemap[]> => [
+    new RNSourcemap(this.bundle!, this.sourcemap!),
+  ]
+
+  private getPayloadsToUpload = async (useGit: boolean): Promise<RNSourcemap[]> => {
+    const payloads = await this.getMatchingRNSourcemapFiles()
+    if (!useGit) {
+      return payloads
+    }
+
+    await this.addRepositoryDataToPayloads(payloads)
+
+    return payloads
+  }
+
+  // GetRepositoryPayload generates the repository payload for a specific sourcemap.
+  // It specifically looks for the list of tracked files that are associated to the source paths
+  // declared inside the sourcemap.
+  private getRepositoryPayload = (repositoryData: RepositoryData, sourcemapPath: string): string | undefined => {
+    let repositoryPayload: string | undefined
+    try {
+      const files = repositoryData.trackedFilesMatcher.matchSourcemap(this.context.stdout, sourcemapPath)
+      if (files) {
+        repositoryPayload = JSON.stringify({
+          data: [
+            {
+              files,
+              hash: repositoryData.hash,
+              repository_url: repositoryData.remote,
+            },
+          ],
+          // Make sure to update the version if the format of the JSON payloads changes in any way.
+          version: 1,
+        })
+      }
+
+      return repositoryPayload
+    } catch (error) {
+      this.context.stdout.write(renderGitDataNotAttachedWarning(sourcemapPath, error.message))
+
+      return undefined
+    }
+  }
+
+  private getRequestBuilder(): RequestBuilder {
+    if (!this.config.apiKey) {
+      throw new InvalidConfigurationError(`Missing ${chalk.bold('DATADOG_API_KEY')} in your environment.`)
+    }
+
+    return getRequestBuilder({
+      apiKey: this.config.apiKey!,
+      baseUrl: getBaseSourcemapIntakeUrl(),
+      headers: new Map([
+        ['DD-EVP-ORIGIN', 'datadog-ci react-native'],
+        ['DD-EVP-ORIGIN-VERSION', this.cliVersion],
+      ]),
+      overrideUrl: `v1/input/${this.config.apiKey}`,
+    })
+  }
+
+  private upload(
+    requestBuilder: RequestBuilder,
+    metricsLogger: MetricsLogger,
+    apiKeyValidator: ApiKeyValidator
+  ): (sourcemap: RNSourcemap) => Promise<UploadStatus> {
```

Comment:
Same, this function can probably be simplified.

Reply:
@lraucy to me that's not a big issue as we probably won't need to edit this function. It is an async function since the last `return` returns a Promise. We could break it in multiple chunks,  but that would need a complete refactoring of the `execute` function to be done properly. 
I'd rather do that in a separate PR.